<!--
To change this template, choose Tools | Templates
and open the template in the editor.
-->
<!DOCTYPE html>
<html>
    <head>
        <title>Differential Gene Expression Analysis</title>
        <meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
    </head>
    <body>
        <p style="text-align: center;"><span style="font-size: 14pt;"><strong>Differential Gene Expression Analysis:</strong></span></p>
        <p style="text-align: justify;">The Differential Gene Expression Analysis enables you to automatically collect all necessary count data for an experiment and run a differential gene expression analysis tool of your choice. ReadXplorer offers
            <object CLASSID="java:org.netbeans.modules.javahelp.BrowserDisplayer">
                <param name="content" value="https://www.bioconductor.org/packages/release/bioc/html/DESeq.html">
                <param name="text" value="<html><u>DESeq</u></html>">
                <param name="textFontSize" value="5">
                <param name="textColor" value="blue">
            </object>,
            <object CLASSID="java:org.netbeans.modules.javahelp.BrowserDisplayer">
                <param name="content" value="https://bioconductor.org/packages/release/bioc/html/DESeq2.html">
                <param name="text" value="<html><u>DESeq2</u></html>">
                <param name="textFontSize" value="5">
                <param name="textColor" value="blue">
            </object>,
            <object CLASSID="java:org.netbeans.modules.javahelp.BrowserDisplayer">
                <param name="content" value="https://www.bioconductor.org/packages/release/bioc/html/baySeq.html">
                <param name="text" value="<html><u>baySeq</u></html>">
                <param name="textFontSize" value="5">
                <param name="textColor" value="blue">
            </object> and the Express Test. Additionally, the same wizard offers export of count data tables. Exemplary results are shown in the screenshots below. For further details about the methods, manual and plots generated by the different tools, please visit the respective tool homepage linked above.</p>
        <ol>
            <li style="text-align: justify;">Choose "<span style="text-decoration: underline;">Tools -&gt; Differential Gene Expression Analysis</span>".</li>
            <li style="text-align: justify;">Select the tool for the analysis (DESeq, DESeq2, baySeq or Express Test).</li>
            <li style="text-align: justify;">Setup the analysis design. The different steps are briefly explained on each wizard page and in detail in the corresponding list below.</li>
            <li style="text-align: justify;">General setup:
                <ul>
                    <li style="text-align: justify;">Select the genomic feature types, to take into consideration for the analysis.
                        The other features types are ignored. </li>
                    <li style="text-align: justify;">Set the offset for the start and stop positions of each genomic feature.
                        Very useful to include otherwise ignored reads for genomic features whose annotation starts further downstream
                        of the actual start position observed in the RNA-Seq data. <br /></li>
                    <li style="text-align: justify;">Decide, if the read orientation is taken into account (only reads on the feature
                        strand are taken into consideration), or if all reads from both strands within range of a genomic feature are used.<br /></li>
                    <li style="text-align: justify;">Only for DESeq, DESeq2 and baySeq: Decide, if you want to store the parameters and results
                        generated from the used R tool for further processing.</li>
                </ul>
            </li>
            <li style="text-align: justify;">Select the mapping classes (see
                <i><a href="nbdocs://de.cebitec.readxplorer.ui/de/cebitec/readxplorer/ui/docs/classification-about.html">Read Classification</a></i>)
                to include in the analysis.</li>
            <li style="text-align: justify;">Click "<span style="text-decoration: underline;">Finish</span>" to start the differential gene expression analysis.</li>
        </ol>
        <p style="text-align: justify;"><span style="text-decoration: underline;">Wizard steps DESeq (two condition design)</span>:</p>
        <ol>
            <li style="text-align: justify;">Define the experiment design: standard two condition design or multifactor design (for further information see
                <a href="http://www.bioconductor.org/packages/release/bioc/html/DESeq.html">DESeq manual</a>).</li>
            <li style="text-align: justify;">First select the reference to use in the upper field and afterwards all tracks to include in the analysis in the second field.</li>
            <li style="text-align: justify;">Assign each track to one of the two conditions by using the arrows next to the two condition fields.
                ReadXplorer automatically identifies the best mode to estimate the dispersion values of the data for the given setup (For details see
                <a href="http://www.bioconductor.org/packages/release/bioc/html/DESeq.html">DESeq manual</a>).</li>
            <li style="text-align: justify;">Continue at "4. General setup" in the general instruction list above.</li>
        </ol>
        <p style="text-align: justify;"><span style="text-decoration: underline;">Wizard steps baySeq</span>:</p>
        <ol>
            <li style="text-align: justify;">First select the reference to use in the upper field and afterwards all tracks to include in the analysis in the second field.</li>
            <li style="text-align: justify;">Define replicate structure: At first select all tracks, which are replicates of the first condition. Then click "<span style="text-decoration: underline;">Add as replicates</span>". Continue with the next conditions in the same way until all tracks are assigned into the replicate structure. If only one track is available per condition, just select one per condition.</li>
            <li style="text-align: justify;">Create baySeq models (for further details see baySeq manual): Similarly to step 2, first select all tracks belonging to your first experimental model, then click the arrow button and continue with all other tracks in the same way until all tracks are assigned. Then click "<span style="text-decoration: underline;">Add group</span>". Now you can create another model. A group containing all tracks is created automatically.</li>
            <li style="text-align: justify;">Continue at "4. General setup" in the general instruction list above.</li>
        </ol>
        <p style="text-align: justify;"><span style="text-decoration: underline;">Wizard steps Express Test</span>:</p>
        <ol>
            <li style="text-align: justify;">First select the reference to use in the upper field and afterwards all tracks to include in the analysis in the second field.</li>
            <li style="text-align: justify;">Assign each track to one of the two conditions by using the arrows next to the two condition fields.</li>
            <li style="text-align: justify;">First continue at "4. General setup" in the general instruction list above. Afterwards continue here.</li>
            <li style="text-align: justify;">Select normalization of the read count values: The normalized read count can be calculated automatically across all genes or house keeping genes can be selected for normalization (hold CTRL while clicking a gene).</li>
            <li style="text-align: justify;">Continue at step 5 in the general instruction list above.</li>
        </ol>
        <p style="text-align: center;"><img style="display: block; margin-left: auto; margin-right: auto;" src="nbdocs://de.cebitec.readxplorer.transcriptionanalyses/de/cebitec/readxplorer/transcriptionanalyses/docs/DifferentialExpressionAnalysis-DESeq-Plot.jpg"/></p>
        <p style="text-align: justify;">This example shows a differential gene expression result using DESeq. The result table shows all values calculated by DESeq for all genomic features of interest. The table can be sorted by the most significantly differential, up- or downregulated genes. The "<span style="text-decoration: underline;">Save Table</span>" button stores the table as an Excel sheet. The "<span style="text-decoration: underline;">Show processing log</span>" button opens another tab with the log of the DESeq run. The&nbsp;"<span style="text-decoration: underline;">Create Graphics</span>" button opens the graphics tab, which allows to generate, view and store different plots, depending on the selected analysis tool. Here, an interactive (denoted by the "i"-icon at the bottom left) M/A plot of the results is shown. All interactive plots allow centering a gene of interest in all synchronized viewers by simply clicking on it. This function enables direct close inspection of the actual underlying data. Each plot can also be saved for further use by clicking "<span style="text-decoration: underline;">Save...</span>".</p>
        <p>
            <b>A brief explanation of the different plots:</b>
        </p>
        <ol>
            <li style="text-align: justify;"><b>DESeq:</b>
                <ul>
                    <li style="text-align: justify;">M/A-plot: An interactive M/A plot similar to DESeq's M/A plot (Log 2 fold change vs. base means)
                        created by ReadXplorer. On the x-axis, the plot visualizes A (normalized mean expression (0.5 * log2 (baseMeanA * baseMeanB))),
                        while the y-axis plots M (log2 fold change). Each dot in the plot can be hovered and selected by left clicking. This action
                        centers the synchronized viewers on the selected gene of interest, enabling visual inspection of the underlying data. A gene
                        is more likely to be differentially expressed, if it is located farther away from 0 on the y-axsis and has a higher normalized mean expression.</li>
                    <li style="text-align: justify;">Gene dispersion vs. normalized mean expression: This plot contains the empirical per gene dispersion
                        values (black dots) on the y-axis and the corresponding fitted dispersion values as a red line plotted against the mean expression
                        strength of each gene on the x-axis. The plot is doubly logarithmic. For details on gene dispersion please read the DESeq
                        publication and the DESeq manual.</li>
                    <li style="text-align: justify;">Log 2 fold change vs. base means: DESeq's M/A plot. It contains normalized mean expression (A)
                        against log2 fold change (M) for each gene.</li>
                    <li style="text-align: justify;">Histogram of p-values: This plot visualizes the probability of genes not to be differentially
                        expressed against its frequency in the experiment.</li>
                </ul>
            <li style="text-align: justify;"><b>DESeq2:</b>
                <ul>
                    <li style="text-align: justify;">M/A-plot: See the description above for DESeq.</li>
                    <li style="text-align: justify;">Per gene estimates against normalized mean expression: This plot contains the empirical per gene dispersion estimates (black dots) on the y-axis, the corresponding fitted dispersion values as a red line and the final maximum a posteriori estimates used in testing (blue) plotted against the mean expression strength of each gene on the x-axis. The plot is doubly logarithmic. For details on gene dispersion please read the DESeq2 publication and the DESeq2 manual.</li>
                    <li style="text-align: justify;">Histogram of p-values: This plot visualizes the probability of genes not to be differentially expressed against its frequency in the experiment.</li>
                </ul>
            <li style="text-align: justify;"><b>baySeq:</b>
                <ul>
                    <li style="text-align: justify;">Priors: The log prior probability means (see Prior probability on Wikipedia) for a selected sample group.</li>
                    <li style="text-align: justify;">M/A plot for count data: Where the log-ratio would be infinite, because the data in
                        one of the sample groups consists entirely of zeros, we plot instead the log-values of the other group.</li>
                    <li style="text-align: justify;">Posterior likelihoods of differential expression against log-ratio: Posterior likelihoods
                        of differential gene expression against log-ratio (where this would be non-infinite) or log values (where all
                        data in the other sample group consists of zeros).</li>
                </ul>
            <li style="text-align: justify;"><b>Express Test:</b>
                <ul>
                    <li style="text-align: justify;">M/A plot: Shows the same data as the interactive M/A-plot of DESeq.</li>
                    <li style="text-align: justify;">Ratio plots: These two plots visualize the gene ratio of AB or BA on the x-axis, while the
                        y-axis shows the calculated confidence value. A higher confidence (top) depicts a higher probability, that the result is trustworthy.</li>
                </ul>
        </ol>
    </body>
</html>
