<!--/
 Copyright (C) 2014 Institute for Bioinformatics and Systems Biology, University Giessen, Germany

 This program is free software: you can redistribute it and/or modify
 it under the terms of the GNU General Public License as published by
 the Free Software Foundation, either version 3 of the License, or
 (at your option) any later version.

 This program is distributed in the hope that it will be useful,
 but WITHOUT ANY WARRANTY; without even the implied warranty of
 MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 GNU General Public License for more details.

 You should have received a copy of the GNU General Public License
 along with this program.  If not, see <http://www.gnu.org/licenses/>.
/-->
<!DOCTYPE html>
<html>
    <head>
        <title>Read Count & Normalization (TPM & RPKM) Calculations</title>
        <meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
    </head>
    <body>
        <p style="text-align: center;"><span style="font-size: 14pt;"><strong>Read Count & Normalization (TPM & RPKM) Calculations:</strong></span></p>
        <p style="text-align: justify;">The read count normalization analysis calculates three values:</p>
        <ul>
            <li style="text-align: justify;"><b>TPM = Transcripts per million</b> according to:
                <object CLASSID="java:org.netbeans.modules.javahelp.BrowserDisplayer">
                        <param name="content" value="http://www.ncbi.nlm.nih.gov/pubmed/20022975">
                        <param name="text" value="<html><u>Li et al. 2010</u></html>">
                        <param name="textFontSize" value="5">
                        <param name="textColor" value="blue">
                    </object>
                <br>The formula is:
                <br><b>TPM = 10^6 * (c / l) * (1 / (sum_i (c_i / l_i)))</b> where
                <br>c = number of mappable reads for gene (or genomic feature)
                <br>l = effective length (or length) of gene (or genomic feature)
                <br>i = 1 - #genes (or genomic features of the same type)
                <br>
            </li>
            <li style="text-align: justify;"><b>RPKM = Reads per kilobase per million mapped reads</b> according to:
                <object CLASSID="java:org.netbeans.modules.javahelp.BrowserDisplayer">
                        <param name="content" value="http://www.ncbi.nlm.nih.gov/pubmed/18516045">
                        <param name="text" value="<html><u>Mortazavi et al. 2008</u></html>">
                        <param name="textFontSize" value="5">
                        <param name="textColor" value="blue">
                </object>
                <br>The formula is:
                <br><b>RPKM = 10^9 * C / (N * L)</b> where
                <br>C = number of mappable reads for feature (e.g. gene)
                <br>N = total number of mappable reads for experiment/data set
                <br>L = sum of feature (e.g. gene) base pairs
                <br>
            </li>
            <li style="text-align: justify;">Raw read count values: All values overlapping the feature and used for the formula calculations.</li>
        </ul>
        <p>for all annotated features of the user-selected feature types (e.g. CDS or gene), which have a raw read count within the given range by:</p>
        <p style="text-align: justify;">1. The <span style="text-decoration: underline;">minimum number of reads</span> (default = 1 read)<br /> 2. The <span style="text-decoration: underline;">maximum number of reads</span> (default = 100000000 reads)<br /> overlapping with a feature.</p>
        <p style="text-align: justify;">When the <b>effective length</b> option is chosen, the length of each genomic feature is shortened by the mean read length of reads mapping to that feature. Therefore, the effective length represents the number of bases of the feature, where reads can start. Mainly used for eukaryotes. The formula for the effective length is as defined by: <object CLASSID="java:org.netbeans.modules.javahelp.BrowserDisplayer">
                        <param name="content" value="http://www.ncbi.nlm.nih.gov/pubmed/20436464">
                        <param name="text" value="<html><u>Trapnell et al. 2010</u></html>">
                        <param name="textFontSize" value="5">
                        <param name="textColor" value="blue"></object><br />
                        <b>L_i = sum_x (λ_F (x)) * l_i - x + 1</b>,<br />
where x is one of the observed read length values in transcript i of length l_i, and λ_F is the fraction of reads for i with length x.
</p>
        <p style="text-align: justify;">The implemented <b>assignment model for reads</b> to features counts a read for a genomic feature if it at least overlaps with the feature. If a read overlaps multiple features:</p>
        <ul>
            <li style="text-align: justify;">If they are sub-features of a larger feature (i.e. CDS and exons of the same gene), they are counted once per gene</li>
            <li style="text-align: justify;">If the overlapping features do not belong together (e.g. a read overlaps two neighboring genes), the read is counted for the first of these features on the respective strand (on the reverse strand this is the last, instead of the first feature).</li>
            <li style="text-align: justify;">If the read is completely contained in multiple features of the same feature type (e.g. for overlapping genes), the fraction of the read is counted for both features (a read completely contained in two genes is thus counted with 0.5 for each of the genes).</li></ul>
        <p style="text-align: justify;">This model including the fraction has been chosen, because according to the literature (Li et al. 2010), models including read mappings overlapping and contained in multiple features show a better correlation to microarray data than models neglecting such read mappings - thus throwing away real data.</p>
        <p style="text-align: justify;"><span style="text-decoration: underline;">Wizard steps:</span></p>
        <ol>
            <li style="text-align: justify;">Click the corresponding toolbar icon (see <i><a href="nbdocs://de.cebitec.readxplorer.ui/de/cebitec/readxplorer/ui/docs/view-about.html">Main Window a)</a></i>) or choose "<span style="text-decoration: underline;">Tools -&gt; Transcription Analyses</span>".</li>
            <li style="text-align: justify;">Select the track(s) to analyze with a left click.</li>
            <li style="text-align: justify;">Select the "<span style="text-decoration: underline;">Read Count & Normalization calculation</span>".</li>
            <li style="text-align: justify;">Select the mapping classes (see <i><a href="nbdocs://de.cebitec.readxplorer.ui/de/cebitec/readxplorer/ui/docs/classification-about.html">Read Classification</a></i>) to include in the analysis. </li>
            <li style="text-align: justify;">Select the Normalization and Read Count Calculation parameters.<br /></li>
            <li style="text-align: justify;">Select the feature types, for which the Read Count & Normalization Calculation shall be carried out and configure an offset (if desired) for both feature boundaries - increasing the number of bases of a feature. NOTE for eukaryotes: If e.g. "Gene" is selected, but not "Exon" the TPM, RPKM and Read Count values still correspond to the exons of that gene and NOT to all reads mapped within the gene's boundaries. If "Exon" or "CDS" are selected e.g. together with "Gene", it means that an extra row for each "Exon" or "CDS" satisfying the given parameters will be shown in the result next to the "Gene" result row. If no "Exon" information is available for a "Gene" it is screened for mRNA/rRNA/tRNA, and only if they are also not present, all reads mapped within the gene's boundaries are counted.<br /></li>
            <li style="text-align: justify;">Click "<span style="text-decoration: underline;">Finish</span>" to start the <span>Read Count & Normalization (TPM & RPKM) Calculation.</span><br /></li>
        </ol>
        <p style="text-align: center;"><img style="display: block; margin-left: auto; margin-right: auto;" src="nbdocs://de.cebitec.readxplorer.transcriptionanalyses/de/cebitec/readxplorer/transcriptionanalyses/docs/TPM-Histogram.png" /></p>
        <p style="text-align: justify;">The screenshot shows an Read Count & Normalization Calculation result with calculated TPM, RPKM and Read Count values in the last three columns of the table. If the <b>effective length</b> option is enabled, this length is shown in the fourth last column. It represents the calculated length of a feature, at which reads may start. This is estimated in connection with the mean read length observed for the feature reads (see above). When the effective length option is disabled, "-1" is printed in this column and "-" is printed in exported tables.  The selected gene and its coverage is centered on the left hand side. The histogram shows the distribution of TPM values on the number of genes (log scale). The tooltip displays the actual count of features, which belong to a histogram bar. The histogram can be switched between TPM and RPKM values by the "<span style="text-decoration: underline;">Switch to ...</span>"-button at the top.</p>
        <p style="text-align: justify;">The <b>total number of assignable mappings</b> in the analysis statistics corresponds to all mappings which could be assigned to any of the selected feature types. Thus when multiple feature types (e.g. genes and CDS) are selected in one analysis, this value can exceed the number of mappings in the data set. This is due to the fact that feature types on different hierarchy levels (like genes and CDS) are treated separately - each read mapping can be assigned to both a gene and a CDS during such an analysis.</p>
    </body>
</html>
