<!--
To change this template, choose Tools | Templates
and open the template in the editor.
-->
<!DOCTYPE html>
<html>
    <head>
        <title>Operon Detection</title>
        <meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
    </head>
    <body>
        <p style="text-align: center;"><span style="font-size: 14pt;"><strong>Operon Detection:</strong></span></p>
        <p style="text-align: justify;">The <span>Operon detection</span> allows to predict operons for prokaryotic genomes on the basis of:<br />
            A) <span style="text-decoration: underline;">Minimal number of spanning reads</span> (default = 5 reads)<br /> between two annotated genes on the same strand. Reads are called "spanning", if they overlap both neighboring features.</p>
        <p style="text-align: center;"><img style="display: block; margin-left: auto; margin-right: auto;" src="nbdocs://de.cebitec.readxplorer.transcriptionanalyses/de/cebitec/readxplorer/transcriptionanalyses/docs/OperonDetection.jpg" alt="operon-detection-result" /></p>
        <p style="text-align: justify;">This is an exemplary result of an operon detection on <em>Corynebacterium glutamicum</em>. The highlighted operon is centered in the Reference and PersistentTrack Viewers. Here, we can observe, that the coverage continues through all genes from cg0064 to cg0055 on the reverse strand. Therefore, we can presume that they are transcribed in an operon. The table lists each neigboring pair of annotations in one row as "<span style="text-decoration: underline;">Feature 1</span>" and "<span style="text-decoration: underline;">Feature 2</span>". In the last four columns we get the information how many reads overlap, span or are positioned between these two annotations.<br /> </p>
    </body>
</html>
